File: validator/ingest_check.py
"""Verify NR's ingest pipeline stored the correct values (esp. counter deltas).

The collector emits *cumulative* counters; New Relic delta-converts them at ingest.
By the telescoping property, sum(metric) over (t0, tN] in NRDB must equal the
emitted cumulative end-to-end difference  C(tN) - C(t0).  Gauges are stored as-is,
so latest(metric) must equal the emitted latest.

We therefore read the emitted series (first/last point) from the OTLP-JSON file,
ask NRDB for the corresponding aggregate, and compare.
"""

from __future__ import annotations

from dataclasses import dataclass

from . import nrql_probe
from .comparator import _within_tolerance
from .comparator_status import (
    INGEST_ERROR, INGEST_MISMATCH, INGEST_NO_DATA, INGEST_OK, INGEST_SKIPPED,
)
from .config import Config
from .ingest_reader import Series
from .metric_map import COMPUTED_SKIP, GAUGE, SUM, value_type_of


@dataclass
class IngestResult:
    metric: str
    attrs: dict
    status: str
    expected: float | None   # from emitted OTLP (delta for sums, latest for gauges)
    actual: float | None     # from NRDB via NRQL
    rel_diff: float | None
    value_type: str | None
    nrql: str = ""
    note: str = ""


def _check_one(series: Series, cfg: Config, client: nrql_probe.NrqlClient) -> IngestResult:
    metric, attrs = series.metric, series.attrs
    vt = value_type_of(metric)

    if metric in COMPUTED_SKIP or vt is None:
        return IngestResult(metric, attrs, INGEST_SKIPPED, None, None, None, vt,
                            note="not a directly-mapped Phase-1 metric")

    since_ms = nrql_probe.to_ms(series.first_ts)
    until_ms = nrql_probe.to_ms(series.last_ts)

    if vt == SUM:
        # Need at least two distinct points spanning time to form a delta.
        if series.n_points < 2 or until_ms <= since_ms:
            return IngestResult(metric, attrs, INGEST_SKIPPED, None, None, None, vt,
                                note="need >=2 emitted points across time for a delta")
        expected = series.last_value - series.first_value
        # Exclude the first point's own delta (it belongs to the prior interval).
        nrql = nrql_probe.build_delta_query(metric, attrs, since_ms + 1, until_ms)
    else:  # gauge
        expected = series.last_value
        nrql = nrql_probe.build_latest_query(metric, attrs, since_ms, until_ms + 1000)

    actual, err = client.run(nrql)
    if err is not None:
        return IngestResult(metric, attrs, INGEST_ERROR, expected, None, None, vt, nrql=nrql, note=err)
    if actual is None:
        return IngestResult(metric, attrs, INGEST_NO_DATA, expected, None, None, vt, nrql=nrql,
                            note="NRQL returned no data (ingest lag, or attrs/window mismatch)")

    ok, rel = _within_tolerance(expected, actual, vt, cfg)
    return IngestResult(metric, attrs, INGEST_OK if ok else INGEST_MISMATCH,
                        expected, actual, rel, vt, nrql=nrql)


def check_ingest(series_map: dict[tuple, Series], cfg: Config) -> list[IngestResult]:
    client = nrql_probe.NrqlClient(cfg)
    results = [_check_one(s, cfg, client) for s in series_map.values()]
    results.sort(key=lambda r: (r.status, r.metric))
    return results


def summarize(results: list[IngestResult]) -> dict[str, int]:
    counts: dict[str, int] = {}
    for r in results:
        counts[r.status] = counts.get(r.status, 0) + 1
    return counts


def has_failures(results: list[IngestResult]) -> bool:
    return any(r.status == INGEST_MISMATCH for r in results)

File: internal/ingestcheck/ingestcheck.go
// Package ingestcheck verifies NR's ingest pipeline stored the correct values.
//
// The collector emits cumulative counters; New Relic delta-converts them at
// ingest. By the telescoping property, sum(metric) over (t0, tN] in NRDB must
// equal the emitted cumulative end-to-end difference C(tN)-C(t0). Gauges are
// stored as-is, so latest(metric) must equal the emitted latest.
package ingestcheck

import (
	"sort"

	"github.com/newrelic-forks/otel-metric-validator/internal/compare"
	"github.com/newrelic-forks/otel-metric-validator/internal/config"
	"github.com/newrelic-forks/otel-metric-validator/internal/ingest"
	"github.com/newrelic-forks/otel-metric-validator/internal/metricmap"
	"github.com/newrelic-forks/otel-metric-validator/internal/nrql"
)

// Ingest-check statuses.
const (
	OK       = "INGEST_OK"
	Mismatch = "INGEST_MISMATCH"
	NoData   = "INGEST_NO_DATA"
	Error    = "INGEST_ERROR"
	Skipped  = "INGEST_SKIPPED"
)

// Result is one ingest-check row.
type Result struct {
	Metric    string            `json:"metric"`
	Attrs     map[string]string `json:"attrs"`
	Status    string            `json:"status"`
	Expected  *float64          `json:"expected"`
	Actual    *float64          `json:"actual"`
	RelDiff   *float64          `json:"rel_diff"`
	ValueType string            `json:"value_type"`
	NRQL      string            `json:"nrql"`
	Note      string            `json:"note"`
}

type runner interface {
	Run(nrqlQuery string) (*float64, error)
}

func checkOne(s ingest.Series, cfg config.Config, client runner) Result {
	vt, mapped := metricmap.ValueTypeOf(s.Metric)
	if metricmap.ComputedSkip[s.Metric] || !mapped {
		return Result{Metric: s.Metric, Attrs: s.Attrs, Status: Skipped, ValueType: vt,
			Note: "not a directly-mapped Phase-1 metric"}
	}

	sinceMS := nrql.ToMS(s.FirstTS)
	untilMS := nrql.ToMS(s.LastTS)

	var expected float64
	var query string
	if vt == metricmap.SUM {
		if s.NPoints < 2 || untilMS <= sinceMS {
			return Result{Metric: s.Metric, Attrs: s.Attrs, Status: Skipped, ValueType: vt,
				Note: "need >=2 emitted points across time for a delta"}
		}
		expected = s.LastValue - s.FirstValue
		// NR stamps each delta at the END of its interval; NRQL UNTIL is exclusive.
		// SINCE first+1 drops the first point's delta (it belongs to the prior
		// interval); UNTIL last+1 includes the final interval's delta. Net window
		// sum = C(last) - C(first) = expected.
		query = nrql.BuildDeltaQuery(s.Metric, s.Attrs, sinceMS+1, untilMS+1)
	} else {
		expected = s.LastValue
		query = nrql.BuildLatestQuery(s.Metric, s.Attrs, sinceMS, untilMS+1000)
	}

	actual, err := client.Run(query)
	if err != nil {
		return Result{Metric: s.Metric, Attrs: s.Attrs, Status: Error, Expected: &expected,
			ValueType: vt, NRQL: query, Note: err.Error()}
	}
	if actual == nil {
		return Result{Metric: s.Metric, Attrs: s.Attrs, Status: NoData, Expected: &expected,
			ValueType: vt, NRQL: query, Note: "NRQL returned no data (ingest lag, or attrs/window mismatch)"}
	}

	ok, rel := compare.WithinTolerance(expected, *actual, vt, cfg)
	status := OK
	if !ok {
		status = Mismatch
	}
	return Result{Metric: s.Metric, Attrs: s.Attrs, Status: status, Expected: &expected,
		Actual: actual, RelDiff: &rel, ValueType: vt, NRQL: query}
}

// Check runs the ingest check over every emitted series.
func Check(series map[string]ingest.Series, cfg config.Config) []Result {
	client := nrql.New(cfg)
	return checkWith(series, cfg, client)
}

func checkWith(series map[string]ingest.Series, cfg config.Config, client runner) []Result {
	results := make([]Result, 0, len(series))
	for _, s := range series {
		results = append(results, checkOne(s, cfg, client))
	}
	sort.Slice(results, func(i, j int) bool {
		if results[i].Status != results[j].Status {
			return results[i].Status < results[j].Status
		}
		return results[i].Metric < results[j].Metric
	})
	return results
}

// Summarize counts results by status.
func Summarize(results []Result) map[string]int {
	counts := map[string]int{}
	for _, r := range results {
		counts[r.Status]++
	}
	return counts
}

// HasFailures reports whether any ingest result is a real mismatch.
func HasFailures(results []Result) bool {
	for _, r := range results {
		if r.Status == Mismatch {
			return true
		}
	}
	return false
}
